convert2rhel/subscription.py

```diff
@@ -277,28 +287,103 @@ def args(self):
 
         return args
 
-    def __call__(self):
+    @property
+    def connection_opts(self):
         """
-        Run the subscription-manager command.
+        This property is a dbus.Dictionary that contains the connection options for RHSM
+        dbus calls.
 
-        Wrapper for running the subscription-manager command that keeps
-        secrets secure.
+        Set :attr:`server_url` to affect this value.
         """
-        if self.password:
-            loggerinst.debug(
-                "Calling command '%s %s'" % (self.cmd, " ".join(hide_secrets(self.args)))
-            )  # lgtm[py/clear-text-logging-sensitive-data]
-            output, ret_code = utils.run_cmd_in_pty(
-                [self.cmd] + self.args, expect_script=(("[Pp]assword: ", self.password + "\n"),), print_cmd=False
-            )
-        else:
-            # Warning: Currently activation_key can only be specified on the CLI. This is insecure
-            # but there's nothing we can do about it for now. Once subscription-manager issue:
-            # https://issues.redhat.com/browse/ENT-4724 is implemented, we can change both password
-            # and activation_key to use a file-based approach to passing the secrets.
-            output, ret_code = utils.run_subprocess([self.cmd] + self.args, print_cmd=False)
+        connection_opts = {}
 
-        return output, ret_code
+        if self.server_url:
+            url_parts = urllib.parse.urlsplit(self.server_url, allow_fragments=False)
+            connection_opts["host"] = url_parts.hostname
+            if url_parts.port is not None:
+                connection_opts["port"] = url_parts.port
+
+            # Current subscription-manager server-side does not take query and fragments at this
+            # path but include them for future proofing.
+            path = urllib.parse.urlunsplit(("", "", url_parts.path, url_parts.query, url_parts.fragment))
+            connection_opts["handler"] = path
+
+        connection_opts = dbus.Dictionary(connection_opts, signature="ss")
+        return connection_opts
+
+    def __call__(self):
+        """
+        Use dbus to register the system with subscription-manager.
+
+        Status of dbus on various platforms:
+            * RHEL6:
+                * DBUS is present but may not be installed.
+                * Install the dbus package and run /etc/rc.d/init.d/messagebus start
+                * dbus-python-0.83.0 is available
+            * RHEL7:
+                * dbus looks to be installed and run by default
+                * dbus-python-1.1.9 is available
+            * RHEL8:
+            * RHEL9:
```

Comment:
State for RHEL 8 and RHEL 9 is the same? If so, I would group them together like: 

```
RHEL8 & RHEL9:
```

Reply:
Good point.  I didn't have a RHEL8 vm handy when I was filling this comment out.  I checked things just now and RHEL8 and RHEL9 are almost identical.  RHEL8 just has a slightly older version of python3-dbus so I've merged them as you suggested.